nixos/modules/services/networking/firewalld/lib.nix

```diff
@@ -0,0 +1,55 @@
+{ lib }:
+
+let
+  inherit (lib) mkOption;
+  inherit (lib.types)
+    either
+    enum
+    nullOr
+    port
+    submodule
+    ;
+  mkPortOption =
+    {
+      optional ? false,
+    }:
+    mkOption {
+      type =
+        let
+          type = either port (submodule {
```

Comment:
Aside: I was looking at the existing `networking.firewall.allow*PortRanges` options, where it felt like a similar `either` union would make sense there instead. I guess we could just `mkRenamedOptionModule ["allowedUDPPortRanges"] ["allowedUDPPorts"]` at some point, once `allowedUDPPorts` has an `either` type.

Definitely not in scope for this PR, it just triggered my memory seeing you take this approach!